File: Drug-stock predictor/Desktop/Drug-stock-predictor/med-predict/go-backend/internal/services/audit.go
package services

import (
	"encoding/json"

	"med-predict-backend/internal/db"
	"med-predict-backend/internal/models"

	"github.com/google/uuid"
)

// AuditService handles audit logging
type AuditService struct {
	db     *db.Database
	logger *Logger
}

// NewAuditService creates an audit service instance
func NewAuditService(database *db.Database, logger *Logger) *AuditService {
	return &AuditService{
		db:     database,
		logger: logger,
	}
}

// LogAction logs a user action for compliance
func (as *AuditService) LogAction(userID, pharmacyID, action, entityType, entityID, ipAddress string, details interface{}) {
	log := &models.AuditLog{
		ID:         uuid.New().String(),
		UserID:     userID,
		PharmacyID: pharmacyID,
		Action:     action,
		EntityType: entityType,
		EntityID:   entityID,
		IPAddress:  ipAddress,
	}

	// Serialize details to JSON
	if details != nil {
		if data, err := json.Marshal(details); err == nil {
			log.Details = string(data)
		}
	}

	// Log to database - never fail the main operation
	if err := as.db.LogAuditEvent(log); err != nil {
		as.logger.Error("failed to write audit log", "error", err.Error(), "action", action)
	}
}

// Audit action constants
const (
	ActionLoginSuccess    = "LOGIN_SUCCESS"
	ActionLoginFailed     = "LOGIN_FAILED"
	ActionUserCreated     = "USER_CREATED"
	ActionUserUpdated     = "USER_UPDATED"
	ActionStockAdded      = "STOCK_ADDED"
	ActionStockAdjusted   = "STOCK_ADJUSTED"
	ActionBatchSubmitted  = "BATCH_SUBMITTED"
	ActionBatchApproved   = "BATCH_APPROVED"
	ActionBatchRejected   = "BATCH_REJECTED"
	ActionPharmacyCreated = "PHARMACY_CREATED"
)

// Common entity types
const (
	EntityTypeUser     = "USER"
	EntityTypePharmacy = "PHARMACY"
	EntityTypeMedicine = "MEDICINE"
	EntityTypeBatch    = "BATCH"
)

File: Drug-stock predictor/Desktop/Drug-stock-predictor/med-predict/py-backend/app/services/audit.py
from sqlalchemy.orm import Session
from app.models import User, AuditLog
import logging
from datetime import datetime
from typing import Optional, Dict, Any
import uuid

logger = logging.getLogger(__name__)


class AuditService:
    """Service for audit logging"""
    
    def __init__(self, db: Session):
        self.db = db
    
    def log_action(
        self,
        user_id: Optional[str],
        pharmacy_id: Optional[str],
        action: str,
        entity_type: str,
        entity_id: Optional[str],
        ip_address: str,
        changes: Optional[Dict[str, Any]] = None,
    ):
        """Log an action to the audit log"""
        try:
            audit_log = AuditLog(
                id=str(uuid.uuid4()),
                user_id=user_id,
                pharmacy_id=pharmacy_id,
                action=action,
                entity_type=entity_type,
                entity_id=entity_id,
                ip_address=ip_address,
                changes=changes,
            )
            self.db.add(audit_log)
            self.db.commit()
        except Exception as e:
            logger.error(f"Error logging audit action: {str(e)}")
            self.db.rollback()


class LoggerService:
    """Centralized logging service"""
    
    def __init__(self):
        self.logger = logging.getLogger(__name__)
    
    def info(self, message: str, **kwargs):
        self.logger.info(f"{message} {kwargs}")
    
    def warning(self, message: str, **kwargs):
        self.logger.warning(f"{message} {kwargs}")
    
    def error(self, message: str, **kwargs):
        self.logger.error(f"{message} {kwargs}")
    
    def debug(self, message: str, **kwargs):
        self.logger.debug(f"{message} {kwargs}")
